Toroidalprimitive

hervasa2 opened the pull request:
**Added Torus Primitive
Added Rotation support to PartialCircle LinePrimitive. The plot recipe of Torus depends on this.**
We should discuss how to add Rotation support in general. From what I see rotations in the Cone primitive only allow a for a rotation matrix of type RotXYZ, which will rotate in the order Z->X->Y. We need all permutations of this and thus it requires more user input (order choice). We also need to allow a rotation of type AngleAxis (what I used in PartialCircle). 
Examples of the Torus Primitive follow, including usage of the sample function and potential calculations:
**Whole coaxial torus, added to example config files**
![torus_whole](https://user-images.githubusercontent.com/41748838/99576796-6cb55600-29a8-11eb-8d6d-1690c4b52d2b.png)
![torus_whole_plots](https://user-images.githubusercontent.com/41748838/99576790-6b842900-29a8-11eb-97b1-ffc9ff694002.png)
**Coaxial torus, but with a partial thick n contact section**
![torus_section_whole](https://user-images.githubusercontent.com/41748838/99576842-7dfe6280-29a8-11eb-8e23-bf823ab2a0a7.png)
![torus_section_whole_plots](https://user-images.githubusercontent.com/41748838/99576850-8191e980-29a8-11eb-83ee-4731a5650deb.png)
**A bulletized PPC**
![bulletized_PPC](https://user-images.githubusercontent.com/41748838/99577065-ccabfc80-29a8-11eb-9e80-85482f65c447.png)
![bulletized_PPC_plots](https://user-images.githubusercontent.com/41748838/99577075-d03f8380-29a8-11eb-9386-0f8c37f38755.png)

lmh91 commented on examples/coaxial_torus_config.json:
```diff
@@ -0,0 +1,118 @@
+            },
+            "geometry": {
+                "type": "torus",
+                "c": 10.0,
+                "a": {
+                    "from": 2.0,
+                    "to": 5.0
+                },
```
Can you explain me what `c` and `a` is? I expected something like `r1` and `r2`?

hervasa2 replied:
`c`(radius from the center of the hole to the center of the torus tube) and `a`(tube radius) are described in lines 23 and 47 respectively. I reviewed some literature to see what is commonly used and ended up picking the notation from [wolfram](https://mathworld.wolfram.com/Torus.html).  I think we should not use `r1` and `r2` because it could get confused with coordinate `r`

schustermartin replied:
Is that not precisely why we should call it something like `r1` and `r2`  because it belongs to the `r` direction of the grid. I would prefer something expressive, `a`, `b`, and `c` are very generic and you have to consult a manual to find out what to put there.

hervasa2 replied:
`r_1` I guess would but then `r_2` would not. How about `r_tube` and `r_center`?

lmh91 replied:
So in general try to go along with OpenSCAD / OpenCSG / Geant4. 
And I strongly prefer something with `r` at these are radii. 
In the config file we can also allow for different notations. Like also r_center and r_tube.
But please add also a documentation to it on top of the struct, like:
```julia
"""
    # struct Torus{T}

`r_1`: ....
`r_2`: ...
...
"""
struct Torus....
    ....
end
```

I am also suspicious about the `a_interval`?

hervasa2 replied:
`a_interval` would be converted to `r_tube_interval` and `c` to `r_center`

lmh91 replied:
I don't understand how a Torus has a fixed interval in `r` as this interval depends on `z`.
Or am I missing something?

lmh91 replied:
Ah, is it that the torus can have a "hole". So like a tube and not a cylinder?

schustermartin replied:
Yes, I believe that's how it works!
![torus](https://user-images.githubusercontent.com/6920236/99795420-0f4a1200-2b2c-11eb-8c62-4609703658cf.png)

lmh91 reviewed (changes requested):
Very nice work!
But please rename `c` and `a` in the config files to something with `r`.
E.g. `r_torus` and `r_tube`. Or just `r1` and `r2`. In the config file we can support several notations. The internal structure we can change later as it does not affect the general user.

hervasa2 commented:
I've renamed the variables everywhere and also added documentation and general rotational matrix type support in PartialCircle. Also added CoaxialTorus to SSD_Examples

schustermartin reviewed (approved):
Very nice Job :)

lmh91 reviewed (approved):
:+1:
